vpr/SRC/route/route_tree_timing.c

```diff
@@ -479,16 +556,587 @@ void free_route_tree(t_rt_node * rt_node) {
+			prev_edge = edge;
+			edge = edge->next; // advance normally as branch still exists
+		}
+	} while (edge);
+	// entirely pruned, this should never be reached due to propagating illegality up to branch point for paths
+	assert(rt_root->u.child_list != nullptr);
+
+	// the sum of its children and its own capacitance
+	rt_root->C_downstream = C_downstream_children + rr_node[inode].C;
+	return false;
+}
+bool prune_route_tree(t_rt_node* rt_root, float pres_fac, vector<t_rt_node*>& reached_sinks, vector<int>& remaining_targets) {
+
+	/* Prune a skeleton route tree of illegal branches (in terms of congestion)						  *
+	 * This is the top level function to be called with the SOURCE node as root 					  *
+	 * returns whether the entire tree has been pruned 												  *
+	 * 																								  *
+	 * assumes R_upstream is already preset for SOURCE, and skip checking parent switch (index of -1) */
+
+	// SOURCE node should never be congested
+	assert(rt_root != nullptr);
+	assert(rr_node[rt_root->inode].get_occ() <= rr_node[rt_root->inode].get_capacity());
+
+	// prune illegal branches from root
+	auto edge = rt_root->u.child_list;
+	t_linked_rt_edge* prev_edge {nullptr};
+	float C_downstream_branches {0};
+
+	while (edge) {
+		bool pruned {prune_illegal_branches_from_route_tree(edge->child, pres_fac, reached_sinks, remaining_targets, 0)};
+		if (pruned) {
+			find_sinks_of_route_tree(edge->child, remaining_targets);
+			edge = free_route_subtree(rt_root, prev_edge);	// returns next edge
+			// previous edge does not change when the current edge gets cut
+		}
+		else {
+			// child still exists and has calculated C_downstream
+			if (!g_rr_switch_inf[edge->iswitch].buffered)
+				C_downstream_branches += edge->child->C_downstream;
+			prev_edge = edge;
+			edge = edge->next; // advance normally as branch still exists
+		}
+	}
+	// for root this can happen if all child branches are pruned
+	if (!rt_root->u.child_list) {
+		return true;
+	}
+
+	rt_root->C_downstream += C_downstream_branches;
+	return false;	
+}
+
+
+void pathfinder_update_cost_from_route_tree(const t_rt_node* rt_root, int add_or_sub, float pres_fac) {
+
+	/* Like pathfinder_update_one_cost, but works with a route tree instead *  
+	 * found sinks are put in reached_sinks if it is not null 				*/
+
+	assert(rt_root != nullptr);
+
+	t_linked_rt_edge* edge {rt_root->u.child_list};
+	
+	// update every node once, so even do it for sinks and join points once
+	for (;;) {
+		pathfinder_update_single_node_cost(rt_root->inode, add_or_sub, pres_fac);
+
+		// reached a sink
+		if (!edge) {
+			return;
+		}
+		// join point (sibling edges)
+		else if (edge->next) {
+			// recursively update for each of its sibling branches
+			do {
+				pathfinder_update_cost_from_route_tree(edge->child, add_or_sub, pres_fac);
+				edge = edge->next;
+			} while (edge);
+			return;
+		}
+
+		rt_root = edge->child;
+		edge = rt_root->u.child_list;
+	}
+}
+
```

Comment:
Unfortunately the breadth first routing uses the traceback (in extension this method) exclusively, so we can't throw it away if we don't also refactor breadth first routing